### DISCUSSION TITLE
Using Lavalink to search for music by URL

### DISCUSSION THREAD

**Discussion by samjesus8** [DSharpPlus/DSharpPlus#1457] [Support]
I am trying to get this to work, so far I have got it to work with normal search terms but not a direct URL

My aim is to get this play command working via a youtube URL like ">play https://www.youtube.com/watch?v=pWO718iy5mY"
This is my current code for this command and I can't seem to work out what else to change:

```
        [Command("play")]
        public async Task PlayMusicURL(CommandContext ctx, Uri URL)
        {
            var userChannel = ctx.Member.VoiceState.Channel;

            //PRE-EXECUTION CHECKS
            if (ctx.Member.VoiceState == null || ctx.Member.VoiceState.Channel == null) //Checking to see if user is in a VC
            {
                await ctx.RespondAsync("Please enter a Voice channel");
                return;
            }

            var lavaInstance = ctx.Client.GetLavalink(); //Getting instance of LavaLink

            if (!lavaInstance.ConnectedNodes.Any()) //Check if the bot is connected to LavaLink
            {
                await ctx.RespondAsync("Connection is not Established");
                return;
            }

            var node = lavaInstance.ConnectedNodes.Values.First();

            if (userChannel.Type != ChannelType.Voice) //Check if the Bot is in a valid Voice Channel
            {
                await ctx.RespondAsync("Invalid Voice Channel");
                return;
            }

            await node.ConnectAsync(userChannel); //Connect to the channel

            var conn = node.GetGuildConnection(ctx.Member.VoiceState.Guild);

            if (conn == null) //If the Bot failed to connect
            {
                await ctx.RespondAsync("Lavalink failed to connect");
                return;
            }

            var loadResult = await node.Rest.GetTracksAsync(URL); //Search for the track using GetTracksAsync
            if (loadResult.LoadResultType == LavalinkLoadResultType.LoadFailed || loadResult.LoadResultType == LavalinkLoadResultType.NoMatches)
            {
                await ctx.RespondAsync("Failed to find a track with this URL"); //Check to see if the Server failed or if there was no matches
                return;
            }

            var track = loadResult.Tracks.First(); //Get the first entry in the search as it is the most accurate

            await conn.PlayAsync(track); //Play the track

            var MusicMSG = new DiscordMessageBuilder()
                .AddEmbed(new DiscordEmbedBuilder()

                .WithColor(DiscordColor.Azure)
                .WithAuthor("𝕤𝕒𝕞.𝕛𝕖𝕤𝕦𝕤𝟠's Music System")
                .WithTitle("Joined Channel " + userChannel.Name)
                .WithDescription("Now Playing: " + track.Title + "\n" +
                                 "Author: " + track.Author + "\n" +
                                 "URL: " + track.Uri)
                .WithFooter("Please use >pause & >resume and not >play when Pausing and Replaying the track")
                );

            await ctx.RespondAsync(MusicMSG);
        }
```

- I used the parameter Uri for URLs
- Apparently the node.Rest.GetTracksAsync() has an overload for URLs but I don't understand how it works

**Comment by masterdragon301** [comment]
yo tengo esto como play.js, funciona con las url, solo pidela a una ia que los compare y funcionara:const { ActionRowBuilder, ButtonBuilder, ButtonStyle, SlashCommandBuilder, EmbedBuilder } = require('discord.js');

const welcomeShown = new Set();

// ─────────────────────────────────────────────
// Elimina tildes y signos para comparaciones
// "¿Por Qué Será?" → "por que sera"
// ─────────────────────────────────────────────
function stripAccents(text) {
    return text
        .normalize('NFD')
        .replace(/[\u0300-\u036f]/g, '')
        .replace(/[¿?¡!.,'"()[\]]/g, '')
        .toLowerCase()
        .replace(/\s+/g, ' ')
        .trim();
}

// ─────────────────────────────────────────────
// Detecta el tipo de URL
// ─────────────────────────────────────────────
function detectUrlType(query) {
    if (!query.startsWith('http://') && !query.startsWith('https://')) return null;
    if (query.includes('spotify.com/track'))    return 'spotify_track';
    if (query.includes('spotify.com/playlist')) return 'spotify_playlist';
    if (query.includes('spotify.com/album'))    return 'spotify_album';
    if (query.includes('youtube.com') || query.includes('youtu.be')) return 'youtube_url';
    return 'url';
}

// ─────────────────────────────────────────────
// Filtra y puntúa resultados de Lavalink
// Descarta shorts, mixes, previews
// Prioriza versiones oficiales
// ─────────────────────────────────────────────
function pickBestTrack(tracks, query) {
    const stripped = stripAccents(query);

    const scored = tracks
        .filter(t => {
            if (!t.duration) return true;
            const sec = t.duration / 1000;
            return sec >= 60 && sec <= 600; // entre 1 min y 10 min
        })
        .map(t => {
            const title      = stripAccents(t.title || '');
            const author     = stripAccents(t.author || '');
            let score        = 0;

            // Coincidencia de título con la query
            if (title.includes(stripped) || stripped.includes(title)) score += 40;

            // Palabras clave de calidad en el título
            if (/official/.test(title))     score += 20;
            if (/audio/.test(title))        score += 15;
            if (/topic/.test(author))       score += 25; // canal "- Topic" de YouTube Music
            if (/music/.test(author))       score += 10;

            // Penalizar covers, remixes no pedidos, karaoke
            if (/karaoke/.test(title))      score -= 50;
            if (/cover/.test(title))        score -= 30;
            if (/remix/.test(title) && !/remix/.test(stripped)) score -= 20;

            // Duración ideal: 2–5 min
            const sec = (t.duration || 0) / 1000;
            if (sec >= 120 && sec <= 300)   score += 15;

            return { track: t, score };
        })
        .sort((a, b) => b.score - a.score);

    return scored.length > 0 ? scored[0].track : null;
}

// ─────────────────────────────────────────────
// Busca con fallback inteligente:
// 1. spsearch (Spotify via lavasrc)
// 2. ytmsearch (YouTube Music)
// 3. scsearch  (SoundCloud)
// ─────────────────────────────────────────────
async function smartSearch(query, manager, user) {
    const optimized = buildOptimizedQuery(query);
    
    // Usar YouTube Music como primera opción (funciona mejor que spsearch)
    const attempts = [
        { prefix: 'ytmsearch', label: '🎵 YouTube Music', q: `${optimized}` },
        { prefix: 'ytmsearch', label: '🎵 YouTube Music (official)', q: `${optimized} official audio` },
        { prefix: 'spsearch',  label: '🎵 Spotify', q: optimized },
        { prefix: 'ytsearch',  label: '🎵 YouTube', q: `${optimized} cancion` },
    ];

    for (const { prefix, label, q } of attempts) {
        if (q.length < 3) continue;
        
        console.log(`🔍 Intentando [${prefix}]: "${q}"`);

        try {
            const result = await manager.search(`${prefix}:${q}`, { requester: user });

            if (!result?.tracks?.length) {
                console.log(`   ↳ Sin resultados`);
                continue;
            }

            const best = pickBestTrack(result.tracks, q);
            if (!best) {
                console.log(`   ↳ Resultados descartados por filtro`);
                continue;
            }

            console.log(`   ↳ ✅ Encontrado: "${best.title}" (${label})`);
            return { track: best, label };

        } catch (err) {
            console.error(`   ↳ Error en ${prefix}:`, err.message);
        }
    }

    return null;
}

// ─────────────────────────────────────────────
// Formatea ms → "3:45" o "1:02:30"
// ─────────────────────────────────────────────
function formatDuration(ms) {
    const s  = Math.floor(ms / 1000);
    const h  = Math.floor(s / 3600);
    const m  = Math.floor((s % 3600) / 60);
    const sc = s % 60;
    if (h > 0) return `${h}:${String(m).padStart(2, '0')}:${String(sc).padStart(2, '0')}`;
    return `${m}:${String(sc).padStart(2, '0')}`;
}

// ─────────────────────────────────────────────
// Comando principal
// ─────────────────────────────────────────────
module.exports = {
    data: new SlashCommandBuilder()
        .setName('play')
        .setDescription('Reproduce música — pon el nombre, artista o link de Spotify/YouTube')
        .addStringOption(option =>
            option.setName('query')
                .setDescription('Ej: Ojitos Lindos | Bad Bunny - Ojitos Lindos | link de Spotify')
                .setRequired(true)),

    async execute(interaction) {
        const { client, guild, member, channel } = interaction;
        const voiceChannel = member.voice?.channel;

        if (!voiceChannel) {
            return interaction.reply({
                content: '❌ ¡Debes estar en un canal de voz!',
                ephemeral: true
            });
        }

        const rawQuery = interaction.options.getString('query').trim();
        await interaction.deferReply();

        try {
            // ── Crear o recuperar el player ──
            let player = client.manager.players.get(guild.id);
            if (!player) {
                player = await client.manager.createPlayer({
                    guildId: guild.id,
                    voiceId: voiceChannel.id,
                    textId:  channel.id,
                    deaf:    true,
                });
            }

            let addedEmbed;
            const urlType = detectUrlType(rawQuery);

            // ── CASO 1: URL directa (Spotify o YouTube) ──
            if (urlType) {
                const typeLabels = {
                    spotify_track:    '🎵 Spotify (track)',
                    spotify_playlist: '📋 Spotify (playlist)',
                    spotify_album:    '💿 Spotify (álbum)',
                    youtube_url:      '▶️ YouTube (link)',
                    url:              '🔗 URL directa',
                };
                const label  = typeLabels[urlType] || '🔗 URL';
                const result = await client.manager.search(rawQuery, { requester: interaction.user });

                if (!result?.tracks?.length) {
                    const embed = new EmbedBuilder()
                        .setColor(0xFF0000)
                        .setTitle('❌ No se encontró nada')
                        .setDescription(`No pude reproducir ese link.\n\nVerifica que sea un link válido de Spotify o YouTube.`);
                    return interaction.editReply({ embeds: [embed] });
                }

                if (result.type === 'PLAYLIST') {
                    for (const t of result.tracks) player.queue.add(t);
                    addedEmbed = new EmbedBuilder()
                        .setColor(0x1DB954)
                        .setTitle('📋 Playlist añadida')
                        .setDescription(`**${result.playlistName}**\n${result.tracks.length} canciones en la cola`)
                        .setFooter({ text: label });
                } else {
                    const track = result.tracks[0];
                    player.queue.add(track);
                    const pos = player.queue.size;
                    const isPlaying = player.playing && !player.paused;

                    addedEmbed = new EmbedBuilder()
                        .setColor(0x1DB954)
                        .setTitle(isPlaying ? '✅ Añadida a la cola' : '▶️ Reproduciendo ahora')
                        .setDescription(`**${track.title}**`)
                        .addFields(
                            { name: '⏱️ Duración', value: track.duration ? formatDuration(track.duration) : 'En vivo', inline: true },
                            { name: '📍 Cola',     value: `#${pos}`, inline: true },
                            { name: '🔍 Fuente',   value: label,     inline: false }
                        )
                        .setFooter({ text: `Solicitado por ${interaction.user.tag}` });
                }

            // ── CASO 2: Búsqueda por texto con fallback inteligente ──
            } else {
                const found = await smartSearch(rawQuery, client.manager, interaction.user);

                if (!found) {
                    const embed = new EmbedBuilder()
                        .setColor(0xFF0000)
                        .setTitle('❌ Canción no encontrada')
                        .setDescription(
                            `No encontré **${rawQuery}** en ninguna fuente.\n\n` +
                            `**Sugerencias:**\n` +
                            `• Verifica la ortografía\n` +
                            `• Usa el formato: \`artista - canción\`\n` +
                            `• Pega un link directo de Spotify o YouTube\n\n` +
                            `**Ejemplos:**\n` +
                            `• \`/play Ojitos Lindos\`\n` +
                            `• \`/play Bad Bunny - Ojitos Lindos\`\n` +
                            `• \`/play Ella Baila Sola\`\n` +
                            `• \`/play Grupo Frontera - Por Que Sera\``
                        );
                    return interaction.editReply({ embeds: [embed] });
                }

                const { track, label } = found;
                player.queue.add(track);
                const pos       = player.queue.size;
                const isPlaying = player.playing && !player.paused;

                addedEmbed = new EmbedBuilder()
                    .setColor(0x1DB954)
                    .setTitle(isPlaying ? '✅ Añadida a la cola' : '▶️ Reproduciendo ahora')
                    .setDescription(`**${track.title}**`)
                    .addFields(
                        { name: '⏱️ Duración', value: track.duration ? formatDuration(track.duration) : 'En vivo', inline: true },
                        { name: '📍 Cola',     value: `#${pos}`, inline: true },
                        { name: '🔍 Fuente',   value: label,     inline: false }
                    )
                    .setFooter({ text: `Solicitado por ${interaction.user.tag}` });
            }

            await interaction.editReply({ embeds: [addedEmbed] });

            // ── Iniciar reproducción si no está sonando ──
            if (!player.playing && !player.paused) await player.play();

            // ── Panel de control (solo primera vez por servidor) ──
            if (!welcomeShown.has(guild.id)) {
                await showControlPanel(interaction, channel, guild, client);
            }

        } catch (error) {
            console.error('❌ Error en comando play:', error);
            const embed = new EmbedBuilder()
                .setColor(0xFF0000)
                .setTitle('🚨 Error al reproducir')
                .setDescription(`No se pudo reproducir: **${rawQuery}**\n\nError: \`${error.message}\``);
            await interaction.editReply({ embeds: [embed] });
        }
    }
};

// ─────────────────────────────────────────────
// Panel de control con botones
// ─────────────────────────────────────────────
async function showControlPanel(interaction, channel, guild, client) {
    welcomeShown.add(guild.id);

    const welcomeEmbed = new EmbedBuilder()
        .setColor(0x1DB954)
        .setTitle('🎵 ¡Bienvenido al sistema de música de Lumi!')
        .setDescription(
            '**🎯 Cómo buscar música:**\n' +
            '✓ `Ojitos Lindos` — solo el nombre\n' +
            '✓ `Bad Bunny - Ojitos Lindos` — artista primero\n' +
            '✓ `Ojitos Lindos - Bad Bunny` — canción primero\n' +
            '✓ Link directo de Spotify (track, playlist, álbum)\n' +
            '✓ Link de YouTube\n\n' +
            '**Controles:**\n' +
            '⏸️/▶️ **Pausa/Reanudar** · ⏭️ **Saltar** · 🔁 **Loop** · 🛑 **Parar**'
        )
        .setImage('https://i.pinimg.com/originals/c9/34/c1/c934c1520092a7701528a7d484b6f39c.gif')
        .setFooter({ text: 'Lumi Global 🌍 • Solo aparece una vez por sesión' })
        .setTimestamp();

    const row = new ActionRowBuilder().addComponents(
        new ButtonBuilder().setCustomId('p_pause').setLabel('⏸️/▶️').setStyle(ButtonStyle.Primary),
        new ButtonBuilder().setCustomId('p_skip').setLabel('⏭️').setStyle(ButtonStyle.Secondary),
        new ButtonBuilder().setCustomId('p_loop').setLabel('🔁').setStyle(ButtonStyle.Success),
        new ButtonBuilder().setCustomId('p_stop').setLabel('🛑').setStyle(ButtonStyle.Danger)
    );

    const panelMsg = await channel.send({ embeds: [welcomeEmbed], components: [row] });
    const collector = panelMsg.createMessageComponentCollector({ time: 86_400_000 });

    collector.on('collect', async i => {
        const p = client.manager.players.get(guild.id);
        if (!p) return i.reply({ content: '❌ No hay música sonando.', ephemeral: true });

        try {
            if (i.customId === 'p_pause') {
                if (p.paused) {
                    await p.pause(false);
                    await i.reply({ content: '▶️ Música reanudada.', ephemeral: true });
                } else {
                    await p.pause(true);
                    await i.reply({ content: '⏸️ Música pausada.', ephemeral: true });
                }
            }
            else if (i.customId === 'p_skip') {
                await p.skip();
                await i.reply({ content: '⏭️ Saltando canción...', ephemeral: true });
            }
            else if (i.customId === 'p_loop') {
                const modes = ['none', 'track', 'queue'];
                const next  = modes[(modes.indexOf(p.loop) + 1) % modes.length];
                p.setLoop(next);
                const emojis = { none: '➡️ Sin bucle', track: '🔂 Bucle: canción', queue: '🔁 Bucle: cola' };
                await i.reply({ content: emojis[next], ephemeral: true });
            }
            else if (i.customId === 'p_stop') {
                await p.destroy();
                welcomeShown.delete(guild.id);
                await i.reply({ content: '🛑 Música detenida. ¡Hasta luego!', ephemeral: true });
                collector.stop();
                const disabledRow = new ActionRowBuilder().addComponents(
                    new ButtonBuilder().setCustomId('p_pause').setLabel('⏸️/▶️').setStyle(ButtonStyle.Primary).setDisabled(true),
                    new ButtonBuilder().setCustomId('p_skip').setLabel('⏭️').setStyle(ButtonStyle.Secondary).setDisabled(true),
                    new ButtonBuilder().setCustomId('p_loop').setLabel('🔁').setStyle(ButtonStyle.Success).setDisabled(true),
                    new ButtonBuilder().setCustomId('p_stop').setLabel('🛑').setStyle(ButtonStyle.Danger).setDisabled(true)
                );
                await panelMsg.edit({ components: [disabledRow] });
            }
        } catch (err) {
            console.error('Error en botón:', err);
            if (!i.replied) i.reply({ content: '🚨 Error al usar el botón.', ephemeral: true });
        }
    });
}
